kyo-data/shared/src/main/scala/kyo/Render.scala

```diff
@@ -0,0 +1,100 @@
+                builder.append("(")
+                val valIter                         = value.asInstanceOf[Product].productIterator
+                val showIter: Iterator[Render[Any]] = shows.productIterator.asInstanceOf
+                if valIter.hasNext then
+                    builder.append(showIter.next().asText(valIter.next()))
+                    ()
+                while valIter.hasNext do
+                    builder.append(",")
+                    builder.append(showIter.next().asText(valIter.next()))
+                    ()
+                end while
+                builder.append(")")
+                builder.toString()
+            end asText
+        end new
+    end sumRender
+
+    inline given [A](using mir: scala.deriving.Mirror.Of[A]): Render[A] = inline mir match
+        case sumMir: scala.deriving.Mirror.SumOf[?] =>
+            val shows = summonAll[Tuple.Map[sumMir.MirroredElemTypes, Render]]
+            new Render[A]:
+                def asText(value: A): Text =
+                    val caseIndex                 = sumMir.ordinal(value)
+                    val showInstance: Render[Any] = shows.productElement(caseIndex).asInstanceOf
+                    showInstance.asText(value)
+                end asText
+            end new
+        case singMir: scala.deriving.Mirror.Singleton =>
+            val label: String = constValue[singMir.MirroredLabel]
+            new Render[A]:
+                def asText(value: A): Text = label
+        case prodMir: scala.deriving.Mirror.ProductOf[?] => inline erasedValue[A] match
+                case _: Tuple =>
+                    inline erasedValue[prodMir.MirroredElemTypes] match
+                        case _: EmptyTuple =>
+                            new Render[A]:
+                                def asText(value: A): Text = "()"
+                        case _ =>
+                            sumRender[A, prodMir.type]("", prodMir)
+                case _ =>
+                    val label: String = constValue[prodMir.MirroredLabel]
+                    inline erasedValue[prodMir.MirroredElemTypes] match
+                        case _: EmptyTuple =>
+                            new Render[A]:
+                                def asText(value: A): Text = label + "()"
+                        case _ =>
+                            sumRender[A, prodMir.type](label, prodMir)
+                    end match
+
+end Render
+
+sealed trait Rendered:
+    private[kyo] def textValue: Text
+
+object Rendered:
+    given [A](using r: Render[A]): Conversion[A, Rendered] with
+        def apply(a: A): Rendered =
+            new Rendered:
+                private[kyo] def textValue: Text = r.asText(a)
+    end given
+end Rendered
+
+extension (sc: StringContext)
+    def t(args: Rendered*): Text =
+        StringContext.checkLengths(args, sc.parts)
+        val pi         = sc.parts.iterator
+        val ai         = args.iterator
+        var text: Text = pi.next()
```

Comment:
is it possible this initial value won't exist?

Reply:
Apparently not. (Just added a test for empty string.)

I basically copied this from the s-string interpolator implementation.